assets/js/modules/adsense/setup/setup-main.js

```diff
@@ -0,0 +1,293 @@
+ * limitations under the License.
+ */
+
+/**
+ * External dependencies
+ */
+import PropTypes from 'prop-types';
+
+/**
+ * WordPress dependencies
+ */
+import { useEffect, useState } from '@wordpress/element';
+import { __, _x, sprintf } from '@wordpress/i18n';
+
+/**
+ * Internal dependencies
+ */
+import Data from 'googlesitekit-data';
+import ProgressBar from '../../../components/progress-bar';
+import ErrorText from '../../../components/error-text';
+import { SvgIcon } from '../../../util';
+import { STORE_NAME } from '../datastore/constants';
+import { STORE_NAME as siteStoreName } from '../../../googlesitekit/datastore/site/constants';
+import {
+	ACCOUNT_STATUS_NONE,
+	ACCOUNT_STATUS_MULTIPLE,
+	ACCOUNT_STATUS_DISAPPROVED,
+	ACCOUNT_STATUS_GRAYLISTED,
+	ACCOUNT_STATUS_PENDING,
+	ACCOUNT_STATUS_NO_CLIENT,
+	ACCOUNT_STATUS_APPROVED,
+	SITE_STATUS_NONE,
+	SITE_STATUS_ADDED,
+	determineAccountID,
+	determineClientID,
+	determineAccountStatus,
+	determineSiteStatus,
+} from '../util/status';
+import SetupAccountCreate from './setup-account-create';
+import SetupAccountSelect from './setup-account-select';
+import SetupAccountDisapproved from './setup-account-disapproved';
+import SetupAccountPending from './setup-account-pending';
+import SetupAccountNoClient from './setup-account-no-client';
+import SetupAccountApproved from './setup-account-approved';
+import SetupSiteAdd from './setup-site-add';
+import SetupSiteAdded from './setup-site-added';
+import {
+	AdBlockerWarning,
+} from '../common';
+const { useSelect, useDispatch } = Data;
+
+export default function SetupMain( { finishSetup } ) {
+	// Get settings.
+	const siteURL = useSelect( ( select ) => select( siteStoreName ).getReferenceSiteURL() );
+	const previousAccountID = useSelect( ( select ) => select( STORE_NAME ).getAccountID() );
+	const previousClientID = useSelect( ( select ) => select( STORE_NAME ).getClientID() );
+	const previousAccountStatus = useSelect( ( select ) => select( STORE_NAME ).getAccountStatus() );
+	const previousSiteStatus = useSelect( ( select ) => select( STORE_NAME ).getSiteStatus() );
+	const accountSetupComplete = useSelect( ( select ) => select( STORE_NAME ).getAccountSetupComplete() );
+	const siteSetupComplete = useSelect( ( select ) => select( STORE_NAME ).getSiteSetupComplete() );
+
+	// Check whether a change submission is in progress.
+	const isDoingSubmitChanges = useSelect( ( select ) => select( STORE_NAME ).isDoingSubmitChanges() );
+
+	// Check whether settings differ from server and are valid.
+	const canSubmitChanges = useSelect( ( select ) => select( STORE_NAME ).canSubmitChanges() );
+
+	// Determine account.
+	const accounts = useSelect( ( select ) => select( STORE_NAME ).getAccounts() );
+	const accountID = determineAccountID( {
+		accounts,
+		previousAccountID,
+	} );
+
+	// Determine client.
+	const clients = useSelect( ( select ) => select( STORE_NAME ).getClients( accountID ) );
+	const clientID = determineClientID( {
```

Comment:
Since we set this as the `clientID` right away… should the state handle this initialisation after `getClients` is resolved? Rather than having a component do it, which is less reusable and strikes me as a more fragile.

Reply:
There are very specific logic flows necessary here for this to work as expected, together with the account status detection logic etc. Having it all in this one component here makes most sense I think. It won't be needed anywhere else.